server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFastViewFactory.java

```diff
@@ -0,0 +1,149 @@
+            Helpers.assertOneMessageId(messageResults);
+
+            MessageResult firstMessageResult = messageResults.iterator().next();
+            List<MailboxId> mailboxIds = Helpers.getMailboxIds(messageResults);
+
+            Message mimeMessage = Helpers.parse(firstMessageResult.getFullContent().getInputStream());
+
+            return MessageFastView.builder()
+                .id(firstMessageResult.getMessageId())
+                .mailboxIds(mailboxIds)
+                .blobId(BlobId.of(blobManager.toBlobId(firstMessageResult.getMessageId())))
+                .threadId(firstMessageResult.getMessageId().serialize())
+                .keywords(Helpers.getKeywords(messageResults))
+                .size(firstMessageResult.getSize())
+                .inReplyToMessageId(Helpers.getHeaderValue(mimeMessage, "in-reply-to"))
+                .subject(Strings.nullToEmpty(mimeMessage.getSubject()).trim())
+                .headers(Helpers.toHeaderMap(mimeMessage.getHeader().getFields()))
+                .from(Emailer.firstFromMailboxList(mimeMessage.getFrom()))
+                .to(Emailer.fromAddressList(mimeMessage.getTo()))
+                .cc(Emailer.fromAddressList(mimeMessage.getCc()))
+                .bcc(Emailer.fromAddressList(mimeMessage.getBcc()))
+                .replyTo(Emailer.fromAddressList(mimeMessage.getReplyTo()))
+                .date(Helpers.getDateFromHeaderOrInternalDateOtherwise(mimeMessage, firstMessageResult))
+                .preview(previews.get(firstMessageResult.getMessageId()))
+                .build();
+        }
+    }
+
+    private final BlobManager blobManager;
+    private final MessageIdManager messageIdManager;
+    private final MessagePreviewStore previewStore;
+    private MessageFullViewFactory messageFullViewFactory;
+
+    @Inject
+    @VisibleForTesting
+    MessageFastViewFactory(BlobManager blobManager, MessageIdManager messageIdManager, MessagePreviewStore previewStore,
+                           MessageFullViewFactory messageFullViewFactory) {
+        this.blobManager = blobManager;
+        this.messageIdManager = messageIdManager;
+        this.previewStore = previewStore;
+        this.messageFullViewFactory = messageFullViewFactory;
+    }
+
+    @Override
+    public List<MessageFastView> fromMessageIds(List<MessageId> messageIds, MailboxSession mailboxSession) {
+        return Mono.from(previewStore.retrieve(messageIds))
+            .flatMapMany(previews -> gatherMessageViewsFlux(messageIds, mailboxSession, previews))
+            .sort(Comparator.comparing(view -> messageIds.indexOf(view.getId())))
+            .collect(Guavate.toImmutableList())
+            .block();
+    }
+
+    private Flux<MessageFastView> gatherMessageViewsFlux(List<MessageId> messageIds, MailboxSession mailboxSession, Map<MessageId, Preview> previews) {
+        List<MessageId> withFastPreview = ImmutableList.copyOf(previews.keySet());
```

Comment:
This method could be much simpler if it focused on the business logic instead of the optimization.
I would:
1. split the previews in two streams without copying or collecting anything
2. call the `Flux.merge()` version every time
3. let the downstream methods handle corner cases like "this Stream/Set is empty", I should return early"

Reply:
BTW, the current complexity is not exercised at all by the testsuite

Reply:
I simplified the code, not sure it fit your mind, but simpler and few more tests which I could image by all my imagination